### PR TITLE
Subnet microblocks

### PR DESCRIPTION
**Pull request by obycode**
* Mine microblocks against anchor block N-1 while waiting for block N to confirm
* When block N confirms, create block N+1, including all transactions from microblocks (N-1).x
* Also broadcast N+1 as a microblock off of N, so that we can continue to build off of that sequence before N is confirmed

See #238.

This current implementation currently works but is not complete in the handling of forking.

### CODE REVIEW

**Review by jbencin — COMMENTED** [review]
Still working through this. A few minor comments so far...

File: src/chainstate/stacks/miner.rs [hunk 1]
```diff
@@ -1785,6 +1787,53 @@ impl StacksBlockBuilder {
...
+            test_debug!(
+                "Miner payout to process: {:?}; user payouts: {:?}; parent payout: {:?}",
+                _miner_payout,
+                _user_payouts,
+                _parent_reward
+            );
+        }
+
+        let burn_tip_info = SortitionDB::get_canonical_burn_chain_tip(burn_dbconn.conn())?;
+        let burn_tip_height = burn_tip_info.block_height as u32;
```
[the remark this answers is not in the document]
Comment (obycode) [reply src/chainstate/stacks/miner.rs:1819]: Although `u32` does give us 80 thousand years of 10 minute blocks 😅

File: src/chainstate/stacks/miner.rs [hunk 2]
```diff
@@ -1785,6 +1787,53 @@ impl StacksBlockBuilder {
...
+        burn_dbconn: &'a SortitionDBConn,
+    ) -> Result<MinerEpochInfo<'a>, Error> {
+        debug!(
+            "Miner epoch begin";
+            "miner" => %self.miner_id,
+            "chain_tip" => %format!("{}/{}", self.chain_tip.consensus_hash,
+                                    self.header.parent_block)
+        );
+
+        if let Some((ref _miner_payout, ref _user_payouts, ref _parent_reward)) = self.miner_payouts
```
[the remark this answers is not in the document]
Comment (obycode) [reply src/chainstate/stacks/miner.rs:1808]: Agreed. I like this better.

File: src/chainstate/stacks/miner.rs [hunk 3]
```diff
@@ -2323,6 +2377,106 @@ impl StacksBlockBuilder {
...
+        let (tip_consensus_hash, tip_block_hash, tip_height) = (
+            parent_stacks_header.consensus_hash.clone(),
+            parent_stacks_header.anchored_header.block_hash(),
+            parent_stacks_header.stacks_block_height,
+        );
+
+        debug!(
+            "Build anchored block off of {}/{} height {}",
+            &tip_consensus_hash, &tip_block_hash, tip_height
+        );
```
[the remark this answers is not in the document]
Comment (obycode) [reply src/chainstate/stacks/miner.rs]: Updated in 15e6f156a.

File: src/chainstate/stacks/miner.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (obycode) [reply src/chainstate/stacks/miner.rs]: Updated in https://github.com/hirosystems/stacks-subnets/commit/15e6f156ad04ae644b59bdf969016d1c0f5304ce.

File: src/chainstate/stacks/miner.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (obycode) [reply src/chainstate/stacks/miner.rs]: Updated in https://github.com/hirosystems/stacks-subnets/commit/15e6f156ad04ae644b59bdf969016d1c0f5304ce.

File: core-contracts/contracts/subnet.clar [hunk 4]
```diff
@@ -183,10 +187,11 @@
         (block (buff 32))
         (subnet-block-height uint)
         (target-chain-tip (buff 32))
+        (target-chain-heght uint)
```
Comment (jbencin) [inline core-contracts/contracts/subnet.clar:190]: Misspelled "height"

File: src/chainstate/stacks/miner.rs [hunk 5]
```diff
@@ -791,7 +779,7 @@ impl<'a> StacksMicroblockBuilder<'a> {
         }
     }
 
-    /// NOTE: this is only used in integration tests.
+    /// Mine a microblock with the specified transactions.
     pub fn mine_next_microblock_from_txs(
```
Comment (kantai) [inline src/chainstate/stacks/miner.rs:783]: I think that `mine_next_microblock_from_txs` should take an `event_dispatcher` and pass it on to `make_next_microblock`.

File: core-contracts/contracts/multi-miner.clar [hunk 6]
```diff
@@ -102,15 +102,27 @@
...
           (signer-principals (try! (fold verify-sign-helper signatures (ok { block-hash: block-data-hash, signers: (list) })))))
          ;; check that the caller is a direct caller!
          (asserts! (is-eq tx-sender contract-caller) (err ERR_UNAUTHORIZED_CONTRACT_CALLER))
          ;; check that we have enough signatures
          (try! (check-miners (append (get signers signer-principals) tx-sender)))
          ;; execute the block commit
-         (as-contract (contract-call? .subnet-v2-0-0 commit-block (get block block-data) (get subnet-block-height block-data) (get target-tip block-data) (get withdrawal-root block-data)))
+         (as-contract
+            (contract-call?
+                .subnet-v2-0-0
```
Comment (jbencin) [inline core-contracts/contracts/multi-miner.clar:118]: The subnet contract was changed, so you'll need to update `VERSION` there and update the contract name here

**Review by jbencin — COMMENTED** [review]
Not finished reviewing yet...

File: testnet/stacks-node/src/neon_node.rs [hunk 7]
```diff
@@ -322,6 +294,100 @@ fn mine_one_microblock(
     return Ok(mined_microblock);
 }
 
+/// Mine and broadcast a single microblock, unconditionally.
+fn mine_one_microblock_from_prev(
+    microblock_state: &mut MicroblockMinerState,
+    sortdb: &SortitionDB,
+    chainstate: &mut StacksChainState,
+    prev_microblocks: &Vec<StacksMicroblock>,
+) -> Result<StacksMicroblock, ChainstateError> {
```
Comment (kantai) [inline testnet/stacks-node/src/neon_node.rs:303]: Is it possible to de-duplicate this and `mine_one_microblock` somewhat? I think if you add a generic function with a trait argument like:

```rust
F: FnOnce (&mut StacksMicroblockMiner, &Secp256k1PrivateKey) -> Result<StacksMicroblock, ChainstateError>
```

Then you could probably eliminate any duplication between `mine_one_microblock` and `mine_one_microblock_from_prev` (which maybe should be called something like `mine_coalesced_microblock`?)

**Review by kantai — APPROVED** [review]
This LGTM -- just some pretty superficial comments. I think you could deduplicate some of the code in the `neon_node`, and you should also pass the event dispatcher through the new microblock mining path.

File: core-contracts/contracts/multi-miner.clar (hunk 6 above)
Comment (obycode) [reply core-contracts/contracts/multi-miner.clar:118]: I'm going to do this as part of the release process. Checklist in progress in Notion.

File: core-contracts/contracts/subnet.clar (hunk 4 above)
Comment (obycode) [reply core-contracts/contracts/subnet.clar]: Fixed in 73e1e5168

File: src/chainstate/stacks/miner.rs (hunk 5 above)
Comment (obycode) [reply src/chainstate/stacks/miner.rs:783]: Addressed in 2421d72fa

File: testnet/stacks-node/src/neon_node.rs (hunk 7 above)
Comment (obycode) [reply testnet/stacks-node/src/neon_node.rs]: I like this idea, but when I try it, I'm getting:
```
error[E0658]: use of unstable library feature 'fn_traits'
   --> testnet/stacks-node/src/neon_node.rs:264:30
    |
264 |         let mblock = mine_fn.call_once((microblock_miner, &microblock_state.miner_key))?;
    |                              ^^^^^^^^^
    |
    = note: see issue #29625 <https://github.com/rust-lang/rust/issues/29625> for more information
```
I'm not familiar with this construct in Rust. Am I doing something wrong?

```rust
let mblock = mine_fn.call_once((&mut microblock_miner, &microblock_state.miner_key))?;
```

Comment (obycode) [reply testnet/stacks-node/src/neon_node.rs]: Addressed in d13979af2

File: testnet/stacks-node/src/neon_node.rs [hunk 8]
```diff
@@ -979,10 +1003,98 @@ fn spawn_miner_relayer(
...
+                                    ).expect("BUG: failed to load anchored cost of parent block")
+                                    .ok_or(NetError::NotFoundError).expect("BUG: failed to load anchored cost of parent block");
+
+                                    // Before updating the miner state, we need to grab the unconfirmed microblocks from the parent
+                                    // block, to be packaged into a microblock for this new anchor block.
+                                    let prev_microblocks = if let Some(miner_state) = microblock_miner_state {
+                                        miner_state.unconfirmed_microblocks
+                                    } else {
+                                        vec![]
+                                    };
```
Comment (jbencin) [inline testnet/stacks-node/src/neon_node.rs:1021]: There's a simpler way to write this:

```suggestion
                                    let prev_microblocks = microblock_miner_state
                                        .map(|state| state.unconfirmed_microblocks)
                                        .unwrap_or_default();
```

File: testnet/stacks-node/src/neon_node.rs [hunk 9]
```diff
@@ -979,10 +1003,98 @@ fn spawn_miner_relayer(
...
+                                                // send it off
+                                                if let Err(e) = relayer.broadcast_microblock(
+                                                    &microblock_miner.parent_consensus_hash,
+                                                    &microblock_miner.parent_block_hash,
+                                                    next_microblock.clone(),
+                                                ) {
+                                                    error!(
+                                                        "Failure trying to broadcast microblock {}: {}",
+                                                        microblock_hash, e
+                                                    );
```
Comment (jbencin) [inline testnet/stacks-node/src/neon_node.rs:1090]: Minor style thing:

```suggestion
                                                    error!(
                                                        "Failure trying to broadcast microblock {microblock_hash}: {e}"                                                   
                                                    );
```

File: testnet/stacks-node/src/neon_node.rs [hunk 10]
```diff
@@ -979,10 +1003,98 @@ fn spawn_miner_relayer(
...
+                                                }
+                                            }
+                                            Err(e) => {
+                                                error!("Failed to mine microblock: {}", e);
```
Comment (jbencin) [inline testnet/stacks-node/src/neon_node.rs:1094]: Minor style thing:

```suggestion
                                                error!("Failed to mine microblock: {e}");
```

**Review by jbencin — APPROVED** [review]
I left some comments on style, but I approve

**Pull request merged**